ui: Add option to connect to custom consumer socket

pan- opened the pull request:
Testing changes in traced and related probes from the UI requires rebuilding Android or manually replacing system Perfetto binaries, which is cumbersome.

This commit introduces a new option allowing the UI to connect to an alternative tracing service running on the target system, instead of the default Perfetto daemon.

When the feature flag connectToLocalAbstractTraceboxSocket is enabled, the recording plugin connects via a local abstract socket named "@tracebox_traced_consumer". To use this feature, start tracebox and websocket_bridge with the environment variable
PERFETTO_CONSUMER_SOCK_NAME=@tracebox_traced_consumer set.

--- 

Note:

My team regularly updates `traced` and works primarily with Android devices that have root access but a read-only system partition. Historically, we maintained internal code that automatically pushed tracebox to the device and started it as part of the recording process whenever a new binary was available. While this made distributing "complete" versions of Perfetto convenient internally, it is impractical for external sharing or upstreaming.

Having a straightforward method to connect to an alternative tracing service directly from the UI would significantly ease the workflow for testing UI changes dependent on updates to `traced`. Hence this proposed change.

Regarding the implementation in this PR, I'm not entirely satisfied. Initially, I tried following the approach used in the [QueryPage plugin](https://github.com/google/perfetto/blob/main/ui/src/plugins/dev.perfetto.QueryPage/index.ts#L33). Unfortunately, `adb_tracing_session.ts` cannot reference `index.ts` without creating circular dependencies. Passing the flag through multiple function layers down to `createAdbTracingSession` seemed excessive. To avoid this, I introduced `local_flag.ts` to centrally store the local flag without circular dependencies—but this solution feels out of place. Guidance or suggestions on a cleaner approach would be greatly appreciated.

LalitMaganti commented on ui/src/plugins/dev.perfetto.RecordTraceV2/index.ts:
```diff
@@ -34,11 +34,19 @@ import {TracedWebsocketTargetProvider} from './traced_over_websocket/traced_webs
   private static recordingMgr?: RecordingManager;
 
   static onActivate(app: App) {
+    LOCAL_ABSTRACT_SOCKET_FLAG.flag = app.featureFlags.register({
```
IMO this should be a user setting not a feature flag. Settings are the right place for user controllable behaviour which is not an implementation detail of the UI.

We used to abuse flags for that but now we have settings, it should live there.

primiano commented on ui/src/plugins/dev.perfetto.RecordTraceV2/adb/adb_tracing_session.ts:
```diff
@@ -18,15 +18,37 @@ import {TracingProtocol} from '../tracing_protocol/tracing_protocol';
+  return LOCAL_ABSTRACT_SOCKET_FLAG.flag.get();
+}
+
+function getConsumerSocketType() {
```
do we ever need only one of them?
If not, could these two be folded into a single

getConsumerSocketForAdb() which returns either localabstract:traced.... or localfs:/dev/socket/... ?

pan- replied:
There are two accessors now: 
- `getTracedConsumerSocketAddressForAdb` which returns the address formated to be used with ADB.
- `getTracedConsumerSocketAddress` which returns the address stored in the settings. For any information printed to the user, it should be better to use this name as it matches what has been set in the setting.

primiano commented on ui/src/plugins/dev.perfetto.RecordTraceV2/local_flag.ts:
```diff
@@ -0,0 +1,19 @@
+
+// Flag that indicates if a non-default socket should be used to connect to traced.
+// It is present in this file circular dependencies.
+export const LOCAL_ABSTRACT_SOCKET_FLAG: {flag?: Flag} = {};
```
This is actually quite pragmatic. I'd just move also the registration here for consistency

```
const settings = {
  consumerSocket: ...,
}

function registerRecordingSettings(app: App) {
 app.settings.register(...)
}

getConsumerSockerFlag() {
   // Calling this before registerRecordingSettings() should never
   // happen and should just trigger a crash
   assertExists(settings.consumerSocket). ....
}
```

pan- replied:
I created a `settings` object which can be use to include all the settings of the `RecordTraceV2` plugin. 
It includes the registration as requested.

primiano reviewed (commented):
Great overall, i'm very supportive of this. I have a few minor comments, on top of lalit's one.
Apollogies for the flags vs settings confusion. settings are fairly new and we haven't migrated "flags" over.
I just spoke with @stevegolton, what is going to happen at some point soon
- the majorify of the current "flags" will be migrated under settings
- The few things left that are actual "flags" will be moved UI-wise under settings in a section called "Experimental features"

But what you are doing here is a pure setting.

pan- committed:
ui: Add setting to connect to custom consumer socket

Testing changes in traced and related probes from the UI requires rebuilding
Android or manually replacing system Perfetto binaries, which is cumbersome.

This commit introduces a new UI setting, "Traced socket address", enabling
connection to an alternative tracing service instead of the default Perfetto
daemon. Users can specify a custom socket address (prefix the address with
"@" to indicate the abstract namespace).

To utilize this, launch tracebox and websocket_bridge with the environment
variable "PERFETTO_CONSUMER_SOCK_NAME" set to match the configured UI address.

Implementation note: A settings aggregator has been introduced to centralize
configuration for the record trace plugin. It is defined in a dedicated file
to avoid circular dependencies.

pan- commented:
Thanks for your feedbacks, I wasn't aware of the Setting API. It is absolutely perfect for this use case. 
It is much better to set the address name to connect to rather than be forced to use a pre-defined one. 

I tidy up the setting registration part. It remains in its own file but it is not specific to the socket address. Other settings of the recording plugin can be added here too.

stevegolton commented on ui/src/plugins/dev.perfetto.RecordTraceV2/settings.ts:
```diff
@@ -0,0 +1,63 @@
+  }
+}
+
+export default new RecordTraceV2Settings();
```
Please don't create more globals just to make access more convenient. Global state causes headaches for us. Perhaps you could inject these settings through from RecordingManager.

abhmen01 commented:
Addressed the issue of 'settings' being a global. It is now injected through the RecordingManager. 
Please continue the review.

abhmen01 commented:
Hi @primiano 
This is ready to be reviewed.

stevegolton reviewed (approved):
LGTM, thanks!

stevegolton commented on ui/src/plugins/dev.perfetto.RecordTraceV2/settings.ts:
```diff
@@ -0,0 +1,61 @@
+export class RecordTraceV2Settings {
+  static readonly baseId = 'dev.perfetto.RecordTraceV2#';
+  static readonly socketAddressDescriptor: SettingDescriptor<string> = {
+    id: 'dev.perfetto.RecordTraceV2#socketAddressSetting',
```
Actually, one thing. We typically try not to tie the id of a setting to a particular plugin, which gives us the flexibility to move them around without folks losing their settings. It's likely we'll move this setting in the near future so please could you call it `dev.perfetto.TracedConsumerSocketAddress` or similar.

abhmen01 replied:
Changed to dev.perfetto.TracedConsumerSocketAddress
Thanks
